### ISSUE TITLE
dancing list items

### ISSUE THREAD

**Issue by derWalter** [Zettlr/Zettlr#4602]
### Description

when clicken, marking text, or moving the cursor with the arrow keys and on every keypress, even in other lines,
the lines of the list items start to jump left and right.

see behavior here: https://youtu.be/EFQyN_0LcsI

### Reproducing

watch video above, or make a list like this and enjoy the show:

1. <item>:
	1.1. <item>:
	<text>

2. <item>:
2.1. <item>:
	- <item>
	 - <item>

2.2. <item>:
	- <item>r, 

2.3. <item>:
- <item>
- <item>
- <item>

### Zettlr Version

Stable (most recent version)

### Specify version

3.0.0

### Installation Method

From the Website or GitHub

### Your Platform

- [X] Windows
- [ ] macOS
- [ ] Linux

### Architecture

- [X] Intel 64bit
- [ ] ARM 64bit

### Operating System Version

Windows 11 Pro, Version: 22H2, Build: 22621.2215, Windows Feature Experience Pack 1000.22662.1000.0

**Comment by nathanlesage** [comment]
Fitting music. Could you please try to remove the sub-numbered items (`2.1` etc) and see if that makes it stop? There's a good chance I'll simply have to adapt the regular expression. I didn't see this because I never use sub-numbering, so there's a good chance that's why I missed it.

**Comment by jdnicholas** [comment]
I am also seeing the same "dancing" behaviour when I have inserted some Latex code into a document. 

In the code below, the `\centering` and `\includegraphics` lines move about
```
\begin{figure}[h]
	\centering  
 	\includegraphics{./assets/file}  
 \end{figure}  
```

If I remove the indentation the dancing stops.

**Comment by derWalter** [comment]
what ever runs in the background gets recorded, the rest is rhythm :p 

so, I changed this:

`1. Gesundheitsberufe:
	1.1. Freie Berufe:
	Ärzte, Psychotherapeuten, Psychologen, Diätologen, Mediation auf Liste des Justizministeriums eingetragen

2. (Gesundheitsbezogene) Gewerbe:
2.1. Reglementierte Gewerbe:
	- Ernährungsberater
	 - Mediation nicht auf Liste des Justizministeriums eingetragen (LSB Gewerbeschein (auch möglich eingeschränkt auf Mediation), oder Unternehmensberatung einschl. Unternehmensorganisation)

2.2. Freie Gewerbe:
	- Human-Energetiker, 

3. Neue Selbstständige (nicht geschützte Berufe)
- Seelsorge
- Moderation
- Trainer
`

to this:
`
1. Gesundheitsberufe:Freie Berufe: Ärzte, Psychotherapeuten, Psychologen, Diätologen, Mediation auf Liste des Justizministeriums eingetragen

2. (Gesundheitsbezogene) Gewerbe:Reglementierte Gewerbe: Ernährungsberater Mediation nicht auf Liste des Justizministeriums eingetragen (LSB Gewerbeschein (auch möglich eingeschränkt auf Mediation), oder Unternehmensberatung einschl. Unternehmensorganisation)

3. Freie Gewerbe: Human-Energetiker, 

4. Neue Selbstständige (nicht geschützte Berufe) Seelsorge Moderation Trainer`

no problems there.
as soon as I add a "Tabulator" in front of a list item, the party starts,
which is synonymic with the code @jdnicholas provided.


I use sub items a lot to structure my texts, I also mix list types as you can see in my trial of lists :P 
usually (like in libre writer/ms office/eg) thats not a problem (guess they are xml/html based, so, ofc its no big deal)

Guess there is the need for a markup language with complexity between markup and teX, to get all the table, list and image stuff sorted out :/ maybe you need to form (or adopt if available? you are the pro of this field :P) something like that

**Comment by nathanlesage** [comment]
AAAAHHH there are *tabs* -- could be that I missed those 😬

**Comment by jchalifour** [comment]
Hi I noticed the same issue (v3, using Manjaro Linux and Gnome, up-to-date). It's in list items for me but also happens where I put a tab for example before some text. The jumping happens whether clicking on the text or just using arrow keys to move the cursor around. Changing the "indent using tabs" setting in the preferences doesn't seem to make a difference. I'm attaching a file in case an example would be of use. 
[jumptest_20230923092701.md](https://github.com/Zettlr/Zettlr/files/12586903/jumptest_20230923092701.md)

**Comment by leowill01** [comment]
> AAAAHHH there are _tabs_ -- could be that I missed those 😬

i also use tabs for lists and im having this problem as well but worse.
heres a link to a screen recording of the problems: https://youtu.be/IXBHqbfGnPc 

- the lists arent aligned. i cant even tell which bullets are on the same levels.
- constant dancing/re-rendering whenever the cursor is moved or characters are inserted/deleted
- some bullets are rendered, others arent
- text cursor sometimes is not clearly in between two characters

**Comment by nathanlesage** [comment]
I have the suspicion that the width of the tabs is severely misaligned, because CodeMirror has to render those elements separately anyways. Just for verification: could you please try and replace the tabs of one test list with spaces to verify that the issue is really with tabs?

**Comment by leowill01** [comment]
> I have the suspicion that the width of the tabs [...]

Yes, replacing all tabs with eg 4 spaces fixes the problem.
i dont see any more dancing and all bullets seem to be aligned correctly.
also have not noticed any more rendering issues yet like cursor misaligned from where its actually inserting chars.

is this a relatively easy fix?

**Comment by nathanlesage** [comment]
> Yes, replacing all tabs with eg 4 spaces fixes the problem.

Wonderful, that makes this all much easier.

> is this a relatively easy fix?

I literally have no idea; that's some sort of code where it's really impossible to predict the easiness – could be solved with two lines of code, or it may require higher-order magic paired with a 20 d20 roll.

**Comment by nathanlesage** [comment]
So, I have now dug a little bit and found a relatively simple logical bug. However, this unfortunately does not stop tabbed indentation from "dancing" around. I do not specifically know what is happening, but somehow the measurements that come in from CodeMirror are always a little bit off whenever a span of text that includes a tab character is being measured.

This could be some internal bug in CodeMirror, or an issue where the custom styling Zettlr applies may interfere (I suspect the latter). The only thing I can suggest for now as a workaround is to turn off "indent with tabs" in the editor preferences. This will stop the "jumpy" behavior of the visual indentation with no drawbacks whatsoever.

**Comment by nathanlesage** [comment]
Quick update: I think I just had a potential idea for what is happening:

1. When you use tabs, the idea is that those tabs are variable in width to match up with invisible vertical lines (which are fixed for all lines) in the document.
2. What my plugin does is measure the point on the vertical line that those align with
3. It then applies a **negative `text-indent`** to **pull back** the first line a little bit to align it up with the other wrapped lines
4. This means that, strictly speaking, this **moves the tab character around**.
5. This could explain the "wiggling", since the tab is always snapping to lines different from whatever point my plugin moves the text to.

If that is really what is happening, this is a problem, because I cannot only move around the second, third, etc. lines, I *must* use the `text-indent` of the first line, otherwise I cannot properly indent everything (not using text indent means that the second line doesn't align with the first character of the first line, but rather with the list marker of the first line).

All in all, it seems we have to treat tabs in Markdown documents. I think the easiest way would be simply to disallow that, but I know that this is not an option, so we need to treat the visual indentation differently for tabs and spaces, and if you mix them, we must have the indentation fail, I think.

**Comment by sensologica** [comment]
In lists some "dancing" happens when a bullet character gets temporarily rendered as a dash when the cursor is placed on a list item. Perhaps there is some kind of invisible element that changes its width behind the scenes and pushes things around?

![zettlr_bug_dancing-list-items](https://github.com/Zettlr/Zettlr/assets/119978540/b0823311-8207-4072-bf72-1761c8015a84)

**Comment by nathanlesage** [comment]
> In lists some "dancing" happens when a bullet character gets temporarily rendered as a dash when the cursor is placed on a list item. Perhaps there is some kind of invisible element that changes its width behind the scenes and pushes things around?
> 
> 
> 
> ![zettlr_bug_dancing-list-items](https://github.com/Zettlr/Zettlr/assets/119978540/b0823311-8207-4072-bf72-1761c8015a84)

This small dancing around only happens in those instances as I cannot control the exact width of these elements; I deem that an acceptable amount of dancing around for now!

**Comment by nathanlesage** [comment]
Quick additional observation: When you indent a regular line (that is not some list item) with a tab character, then it somehow DOESN'T dance around and the visual indent works flawlessly.

The more I think about this issue, the more puzzling it becomes.

**Comment by leowill01** [comment]
@nathanlesage this is an ignorant comment bc i know you went through an extensive rewrite for v3, but this wasnt an issue in previous versions of zettlr <v3. any way to re-add the previous code for rendering tabs in lists or is that not an option?

**Comment by nathanlesage** [comment]
> any way to re-add the previous code for rendering tabs in lists or is that not an option?

Well, adding back the previous code would mean nothing less than rolling back everything I've changed for 3.0

Finding the cause of this issue is MUCH less work than that, believe me.

**Comment by 87113** [comment]
Still in the 3.0.2 

```
* test
	* item 

text 

* a list
    + 2023
    	- The last
```
[Capture vidéo du 2023-10-26 10-29-05.webm](https://github.com/Zettlr/Zettlr/assets/102969362/6de4e711-851f-4b61-84d5-cc6d4915446b)

**Comment by nathanlesage** [comment]
@87113 No need to reiterate that it still happens – the issue will be closed only when this issue does no longer occur.

**Comment by learn-more** [comment]
Minimal reproduction:
```

- Norma indented text
	- Indented level two
```

Results in:
![image](https://github.com/Zettlr/Zettlr/assets/610685/c9e6e294-cbed-4361-b5c5-2f704e3b7337)

Note the cursor trailing behind the `o` character, when I press backspace ONCE it will already remove the `o`.

When typing on the first line, the second line bounces around:
![image](https://i.imgur.com/cSp98OD.gif)

The indenting is a `tab` character.
This are my editor settings:
![image](https://github.com/Zettlr/Zettlr/assets/610685/a7fe17c8-30f6-46c8-a945-215c9244903e)

**Comment by learn-more** [comment]
Some notes I gathered while looking into it:
- The 'source' of the jumping items is `view.coordsAtPos` inside the `visual-indent.ts` file.
  - This seems to cycle trough 4 values (in the minimal reproducible sample)
  - This only seems to happen with a 'tab' character in front of the dash, not with 4-spaces
  - It seems to be that the **width** of the actual tab character is the thing that moves around, influencing the text after it.

**Comment by nathanlesage** [comment]
> Some notes I gathered while looking into it:
> 
>     * The 'source' of the jumping items is `view.coordsAtPos` inside the `visual-indent.ts` file.
>       
>       * This seems to cycle trough 4 values (in the minimal reproducible sample)
>       * This only seems to happen with a 'tab' character in front of the dash, not with 4-spaces
>       * It seems to be that the **width** of the actual tab character is the thing that moves around, influencing the text after it.
> 
> 
> Logging added to investigate this in [learn-more/Zettlr@`wip`/4602_logging](https://github.com/learn-more/Zettlr/tree/wip/4602_logging?rgh-link-date=2023-11-01T20%3A46%3A47Z)

Exactly that. I am still thinking about solutions, but I think the easiest way would be to simply not visually indenting when there are tabs used for indentation, because this seems to me an interaction in the layouting process of Chromium where it will first apply one of the two settings (text indent or padding-left) and then immediately update the tab position before applying the other setting (which would cancel out the first one), thereby making the cursor jump.

In other words, I believe we're stepping into weird low-level CSS territory here over which I have no power.

**Comment by gilisaacgabriel** [comment]
I have somewhat the same issue:

https://github.com/Zettlr/Zettlr/assets/94884100/d376ecb5-0a72-4109-b03a-0eae64eec458

**Comment by learn-more** [comment]
> I have somewhat the same issue:
> Screen.Recording.2023-11-09.at.2.26.00.PM.mov

To confirm: If you replace the tabs with spaces, the issue is gone for you?

**Comment by limabravo7** [comment]
> I have somewhat the same issue:
> 
>  Screen.Recording.2023-11-09.at.2.26.00.PM.mov

same behaviour for me, and I realized a connection to the editor settings "Indent using tabs".
When **activated**: jumping problem occurs as shown by [gilisaacgabriel](https://github.com/gilisaacgabriel).

When **deactivated**: editor works as expected with the intended indentation, no jumping (as already mentioned by @nathanlesage). But the TAB is still converted to 4 spaces. (video 1)

https://github.com/Zettlr/Zettlr/assets/90968816/7bbeda2e-b962-49c2-bdf4-7856fa697a9f

The YAML front matter behaves differently when mentioned option is **deactivated**: It does **not** convert to 4 spaces and introduces some extra indentation when starting to type (see video 2. first keyword with 4 manual spaces. 2nd with TAB)

https://github.com/Zettlr/Zettlr/assets/90968816/29f2e04f-66c3-4926-8d48-9e554bb80a6e


Maybe this can help, have a nice day

**Comment by nathanlesage** [comment]
Quick update: I just touched the plugin again, did a few seemingly unrelated problems and, after initial "dancing" of tab items, I couldn't reproduce this anymore. I have literally no idea if that is just a hunch, but feel free to keep an eye out in the coming nightlies if this has miraculously fixed itself, or not.

**Comment by learn-more** [comment]
@nathanlesage This is still an issue for me in 3.1.0-beta.1 .

**Comment by nathanlesage** [comment]
Thank you for confirming!

**Comment by ienzo** [comment]
Hi, I would like to confirm that dancing lists are still a problem in 3.1.1. I'm on Linux.

![Peek 2024-05-06 12-02](https://github.com/Zettlr/Zettlr/assets/4653894/d31aac2e-cb60-4a54-abce-085cd6f8f1f9)

**Comment by leowill01** [comment]
checking back in to say im still seeing this issue on v3.2.0.
any update on what might be causing this?
this is literally the only thing wrong with zettlr for my current uses - if fixed it would be perfect!

**Comment by ienzo** [comment]
This is why Discord is a bad platform for collective knowledge. This has been answered but it's not publicly indexed. 

Go to `Preferences…`>`Preferences…`>`Editor`>`Indent using tabs instead of spaces`

If it is checked, uncheck it. If it is unchecked, check it and then uncheck it. This should fix dancing lists for new lists. It doesn't fix lists that are already dancing, but you can change their indentation with spaces instead of tabs to fix it. If it helps, my indentation size is 4, but I don't think that matters for fixing it.

**Comment by nathanlesage** [comment]
This is merely a workaround -- it works with spaces but not with tabs, but what my personal goal for this issue is is to have it also not happen if people choose to use tabs.

**Comment by leowill01** [comment]
> This is why Discord is a bad platform for collective knowledge. This has been answered but it's not publicly indexed.

not sure why Discord is relevant, but it has not been answered.
as @nathanlesage points out, this issue was identified specifically for when tab indents are used (for those of us that dont indent with spaces).
it was never a problem when indenting lists with spaces.

**Comment by Liam-Twomey** [comment]
This actually explains an issue I was having with jumping tables as well. I thought it was a bug in the table editor, it never occurred to me that this was a tab issue. The problem seems similar to #3488 and #1621, but if I understand correctly it's an unrelated issue.

**Comment by Valdus92** [comment]
This interrupts the beautiful work-flow between Zettlr and Logseq. 

I have come to believe that I think in outlines. 

Thus I usually randomind (brain dump if you will- too much time spent with divination & RPG's) in logseq. 

When I think I have something sufficient to write an essay or a coherent subject- I write it in Zettlr. Since, for now, logseq doesnt really have the place for formal writing at this time. 

But with this epileptic outline, I just can't flow, can't write. So my zettlr is missing out. 

Spaces for tabs would be nice but, in formal writing I want to indent just the first line, not the whole paragraph. 

This issue MUST be resolved.

**Comment by nathanlesage** [comment]
> This issue MUST be resolved.

I agree, but the problem still stands that the way browsers calculate tab sizes seems insanely difficult to deal with. As soon as I get a proper solution for preventing tab characters to randomly change their widths, I'll fix this issue. Until then, I'm afraid you'll have to use spaces instead …

**Comment by kyaso** [comment]
Wow, glad you found this neat solution!!

**Comment by b100dian** [comment]
4 lines of code, 30 lines of comments - really appreciating what are you doing there :+1:

**Comment by ienzo** [comment]
Congrats! Glad to hear it!

**Comment by kethinov** [comment]
> > In lists some "dancing" happens when a bullet character gets temporarily rendered as a dash when the cursor is placed on a list item. Perhaps there is some kind of invisible element that changes its width behind the scenes and pushes things around?
> > ![zettlr_bug_dancing-list-items](https://private-user-images.githubusercontent.com/119978540/270925431-b0823311-8207-4072-bf72-1761c8015a84.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU3NjE1MzcsIm5iZiI6MTczNTc2MTIzNywicGF0aCI6Ii8xMTk5Nzg1NDAvMjcwOTI1NDMxLWIwODIzMzExLTgyMDctNDA3Mi1iZjcyLTE3NjFjODAxNWE4NC5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEwMVQxOTUzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NjA5NDg2OWUxMTBmMzFjMmEwODA4YWE4MTY0ZTQwMTRhYWUyZGI4NGU1ZTc3YWIzNmJmMzQ0YjZmOGFjYTI4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jC_M3IDsORC_2OpTMMl1XKj5ozsFxFdnwsi2kqWEc7E)
> 
> This small dancing around only happens in those instances as I cannot control the exact width of these elements; I deem that an acceptable amount of dancing around for now!

Here is a solution to that "dancing" as well: https://github.com/Zettlr/Zettlr/discussions/5543
